### PR TITLE
UIKit app review

### PR DESCRIPTION
**Pull request by VojtaStavik**

### CODE REVIEW

File: WeatherApp-UIKit/WeatherApp-UIKit/AppDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  // MARK: UISceneSession Lifecycle
+
+  func application(
+    _ application: UIApplication,
+    configurationForConnecting connectingSceneSession: UISceneSession,
+    options: UIScene.ConnectionOptions
+  ) -> UISceneConfiguration {
+    // Called when a new scene session is being created.
+    // Use this method to select a configuration to create the new scene with.
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/AppDelegate.swift:30]: This comments is generate with `Xcode` ? when yes, I think we should remove it

File: WeatherApp-UIKit/WeatherApp-UIKit/AppDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  func application(
+    _ application: UIApplication,
+    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
+  ) -> Bool {
+    // Override point for customization after application launch.
+    return true
+  }
+
+  // MARK: UISceneSession Lifecycle
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/AppDelegate.swift:22]: Prefer format `// MARK: - `

File: WeatherApp-UIKit/WeatherApp-UIKit/LocationProvider.swift [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  }
+}
+
+extension LocationProvider.State: Equatable {
+  static func == (lhs: LocationProvider.State, rhs: LocationProvider.State) -> Bool {
+    switch (lhs, rhs) {
+    case (.loading, .loading): return true
+    case let (.location(l), .location(r)): return l == r
+    case let (.error(l), .error(r)):
+      return l.localizedDescription == r.localizedDescription
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/LocationProvider.swift:38]: IMHO: don't use one letter naming like `r` or `l` especially in presentation

File: WeatherApp-UIKit/WeatherApp-UIKit/LocationProviderTests.swift [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import RxSwift
+import RxTest
+import CoreLocation
+
+class LocationProviderTests: XCTestCase {
+
+  var scheduler: TestScheduler!
+  var disposeBag: DisposeBag!
+
+  var sut: LocationProvider!
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/LocationProviderTests.swift:21]: Can you add comment for explain `sut` ?

File: WeatherApp-UIKit/WeatherApp-UIKit/ViewController.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
+  @IBOutlet weak var errorLabel: UILabel!
+
+  // MARK: Lifecycle
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    viewModel.state
+      .drive(onNext: { [unowned self] (state) in
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/ViewController.swift:38]: ` (state) `  ->  ` state `

File: WeatherApp-UIKit/WeatherApp-UIKit/ViewControllerTests.swift [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    vm.dislikeButtonTapped.bind(to: dislike).disposed(by: disposeBag)
+
+    sut.dislikeButton.sendActions(for: .touchUpInside)
+    sut.likeButton.sendActions(for: .touchUpInside)
+
+    XCTAssertEqual(like.events.count, 1)
+    XCTAssertEqual(dislike.events.count, 1)
+  }
+
+    func testErrorState() {
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/ViewControllerTests.swift:87]: Formatting

File: WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPI.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+          location: $0.name
+        )
+      }
+  }
+
+  private static func createRequest(for coordinates: CLLocationCoordinate2D) -> URLRequest {
+    let baseURL = URL(string: "https://api.openweathermap.org/data/2.5/weather")!
+    var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
+    components.queryItems = [
+      .init(name: "APPID", value: "c24e21ebebee093dc832ddc161134d91"),
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPI.swift:59]: It's safety push  weather `APPID` to public git?

File: WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPI.swift [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  let weather: [Weather]
+
+  let name: String
+
+  struct MainDetails: Codable {
+    let temp: Double
+  }
+  let main: MainDetails
+}
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPI.swift:81]: Can you add `Weather` and `MainDetails` under `WeatherResponseDTO` ?

File: WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPITests.swift [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    return try! Data(contentsOf: url)
+  }()
+}
+
+extension Array where Element == URLQueryItem {
+
+  /// Returns the value of the URLQueryItem with the given name. Returns `nil`
+  /// if the query item doesn't exist.
+  fileprivate subscript(_ name: String) -> String? {
+    self.first(where: { $0.name == name}).flatMap({ $0.value })
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPITests.swift:99]: `{ $0.name == name}` -> `{ $0.name == name }`

Comment (jantimar) [reply WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPITests.swift:99]: Maybe we should add `swiftlint`

File: WeatherApp-UIKit/WeatherApp-UIKit/WeatherViewModelTests.swift [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+      .init(time: 0, value: .next(.location(mockLocations[0])))
+    ])
+
+    let api: WeatherViewModel.API = { _ in .just(.mock) }
+
+    var likedWeather: Weather?
+    var dislikedWeather: Weather?
+
+    let storage: WeatherViewModel.Storage = { _weather, liked in
+      if liked { likedWeather = _weather } else { dislikedWeather = _weather }
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKit/WeatherViewModelTests.swift:135]: can we write it on more lines?

File: WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+//  TestURLProtocol.swift
+//  WeatherApp-UIKitTests
+//
+//  Created by Vojta on 04/01/2020.
+//  Copyright © 2020 Industrial Binaries. All rights reserved.
+//
+
+import Foundation
+
+
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift:11]: 2 empty lines in row, please remove one

File: WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  override class func canInit(with request: URLRequest) -> Bool {
+    guard let url = request.url else { return false }
+    return mockResponses.keys.contains(url.withoutQueryComponents)
+  }
+
+  override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+    // Overriding this function is required by the superclass.
+    return request
+  }
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift:27]: Maybe move it on bottom of class to `stopLoading` when this method is not important of us

File: WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+//  Created by Vojta on 04/01/2020.
+//  Copyright © 2020 Industrial Binaries. All rights reserved.
+//
+
+import Foundation
+
+
+/// The URLProtocol subclass allowing to intercept the network communication
+/// and provide custom mock responses for the given URLs.
+class TestURLProtocol: URLProtocol {
```
Comment (jantimar) [inline WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift:14]: 👍

**Comment by jantimar** [comment]
Please update `README.md`

File: WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift (hunk 12 above)
Comment (VojtaStavik) [reply WeatherApp-UIKit/WeatherApp-UIKitTests/TestURLProtocol.swift:27]: 👍

File: WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPITests.swift (hunk 9 above)
Comment (VojtaStavik) [reply WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPITests.swift:99]: 👍

File: WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPI.swift (hunk 7 above)
Comment (VojtaStavik) [reply WeatherApp-UIKit/WeatherApp-UIKit/WeatherAPI.swift:59]: I think it's OK for now

File: WeatherApp-UIKit/WeatherApp-UIKit/LocationProviderTests.swift (hunk 4 above)
Comment (VojtaStavik) [reply WeatherApp-UIKit/WeatherApp-UIKit/LocationProviderTests.swift:21]: I thought I'd start using for the thing the test is testing. https://stackoverflow.com/questions/7321407/what-is-sut-and-where-did-it-come-from

File: WeatherApp-UIKit/WeatherApp-UIKit/LocationProvider.swift (hunk 3 above)
Comment (VojtaStavik) [reply WeatherApp-UIKit/WeatherApp-UIKit/LocationProvider.swift:38]: I know but `lhs` and `rhs` already exist in the context. It's clear enough I think.

**Comment by VojtaStavik** [comment]
Thank you for your comments @jantimar! I made changes from the majority of your comments. I'm closing this because I rebased everything so `master` history is clear.

**Pull request closed without merging**